### DISCUSSION TITLE
라이브러리

### DISCUSSION THREAD

**Discussion by HyoJeong-Park** [wa-reureu-studyroom/toy-project-dailymusic#27] [CS Rb]
### Babel
최신 자바스크립트 언어를 구형의 브라우저가 이해할 수 있게 바꿔주는 역할
자바스크립트 트랜스파일러로 다른 실행환경에서도 돌아갈 수 있도록 언어는 유지한체 소스 코드의 형태만 바꿔줍니다. 많은 개발자가 바벨을 자바스크립트의 컴파일러로 부르기도 합니다. 바벨을 이용하면 es6 이상의 문법으로 작성한 자바스크립트 코드를 es5 이하 버전의 문법으로 작성한 것처럼 소스 코드 내의 문법 형태를 바꿀 수 있습니다. 바벨을 사용함으로서 최신 문법이 적용되지 않은 환경에서도 바벨로 형태를 바꾸어 제대로 작동하게 도와줄 수 있습니다.

[babel](https://www.daleseo.com/js-babel/)

### Webpack 
우리가 작성한 자바스크립트 파일을 컴파일해서 작은 단위로 만들어서 좀 더 빠르게 작동할 수 있게 해주는 역할
웹 팩은 모듈 번들링(html에 들어가는 여러 자바스크립트 파일을 하나의 파일로 만드는 작업)이라고도 부르는데, 쉽게 말해 필요로 하는 다수의 자바스크립트 파일을 하나의 자바스크립트 파일로 만들어주는 역할을 합니다. 파일을 컴파일할 때 여러 모듈을 불러오는 데에는 시간이 걸립니다. 그래서 여러 파일을 하나의 파일로 번들링하여 웹 페이지의 성능을 최적화할 수 있습니다. 

[webpack](https://velog.io/@yon3115/%ED%94%84%EB%A1%A0%ED%8A%B8%EC%97%94%EB%93%9C-%ED%95%84%EC%88%98-Webpack%EC%9D%B4%EB%9E%80)

### Pollyfill
직역하면 충전 솜이라는 뜻으로, 브라우저에서 지원하지 않는 코드를 사용 가능한 코드 조각이나 플러그인으로 채워주는 역할을 합니다. 예를 들어서 es5 문법을 사용하는 환경이라면 es6의 map, promise 등을 지원하지 않기때문에 이 경우에 pollyfill을 사용하는 것입니다. 그렇게 되면 Map, promise 등을 사용가능한 객체로 바꿔주기에 코드적으로 문제가 발생하지 않고 동작할 수 있습니다.

[Pollyfill](https://velog.io/@katanazero86/polyfill%ED%8F%B4%EB%A6%AC%ED%95%84-%EC%9D%B4%EB%9E%80)

### Vite
기존 번들링 과정에서 프로젝트의 규모 등에 따라 모듈의 개수가 증가하면서 자바스크립트 기반의 도구는 병목 현상이 일어나게 되었습니다. 혹은 개발 서버에서 오랜 시간을 기다려야 한다든지 변경된 파일이 적용되기까지 소요되는 시간이 증가하였습니다. vite는 브라우저에서 지원하는 es modules(ESM) 과 네이티브 언어로 작성된 자바스크립트 도구를 활용하여 번들링하는 라이브러리입니다. 이를 통해 기존 번들러와 비교해 10배에서 100배 정도의 빠른 속도를 지원하여 개발 환경을 보다 개선할 수 있습니다.

(포트폴리는 사진과 영상을 위주로 업로드 하는 서비스를 제공하기 때문에, 변경 파일이 적용되기까지의 소요 시간을 줄이고자 하여 vite 를 사용하였습니다.)

[vite](https://ko.vitejs.dev/guide/why.html)

## Q. Node.js 동작방식
→  BFF
프론트엔드가 연산이 필요한 것은 백엔드에서 연산해서 보내주는 경우가 많음. || 리소스들을 백엔드에 저장 
자바스크립트를 브라우저 밖에서도 실행가능하도록 하는 자바스크립트 런타임입니다. 비동기 이벤트 주도 런타임으로써 확장성 있는 네트워크 애플리케이션을 만들 수 있도록 설계되었습니다.이벤트 리스너에 등록해둔 콜백함수를 실행시키는 방식으로 동작하고 이벤트 루프가 이를 동작하게끔 해주는 방식입니다. node js 는 v8 엔진, 내장 라이브러리, libuv로 구성되어 있습니다. 그리고 node.js는 싱글스레드로서 하나의 스레드가 하나의 요청만을 수행할 수 있다는 특징이 있습니다.

[node.js 1](https://velog.io/@wjdaud107/Node-JS-%EB%8F%99%EC%9E%91%EC%9B%90%EB%A6%AC)
[node.js 2](https://mniyunsu.github.io/node-loop/)

## Q. TypeScript
타입스크립트는 자바스크립트의 슈퍼셋인 오픈 소스 프로그래밍 언어입니다. 클라이언트 사이드와 서버 사이드 모두에 사용할 수 있고 자바스크립트 슈퍼셋이기 때문에 자바스크립트 엔진을 사용합니다. 작성자가 원하는 타입을 정의하고 코드를 짜면, 자바스크립트로 컴파일 되어 실행할 수 있습니다. 타입스크립트를 사용하여 타입 에러로 발생하는 문제를 사전에 방지할 수 있고 미리 타입을 정의하기 때문에 런타임 할 때 타입을 결정하기 위해 걸리는 시간을 줄일 수 있습니다. 그리고 미리 타입을 정의하기 때문에 협업을 용이하게 할 수 있습니다.

[typescript](https://fomaios.tistory.com/entry/TS-%ED%83%80%EC%9E%85%EC%8A%A4%ED%81%AC%EB%A6%BD%ED%8A%B8TypeScript%EB%9E%80)

## Q. (추가로) next.js

**Comment by Emma-Hyejin** [comment]
### 미들웨어 

미들웨어는 라이브러리가 아닌 소프트웨어 아키텍쳐의 일부입니다. 
소프트웨어 시스템에서 컴포넌트 간 통신 및 데이터 처리를 중간에서 처리하는 " 중개자 " 역할을 합니다. 라이브러리를 확장하거나 보완하는 역할을 수행합니다. 

*미들웨어는 라이브러리가 아니다.* 
*하지만 라이브러리 중에서 미들웨어 역할을 하는 라이브러리들이 있다.*

#### 미들웨어가 하는 작업들 
1. 요청과 응답 처리 
2. 로깅 및 디버깅
3. 인증 및 권한 부여
4. 데이터 변환 및 가공
5. 캐싱 및 성능 최적화
6. 데이터 검증 및 유효성 검사
7. 라우팅 및 경로 처리

#### 라이브러리 vs 미들웨어 
라이브러리는 코드 및 기능을 재사용하기 위한 모듈의 집합으로 특정 기능 또는 작업을 수행하기 위해 다른 프로그램에서 import 해서 사용할 수 있습니다. 라이브러리는 프로젝트에서 필요한 기능을 확장하고 향상시킬 수 있습니다. 
- Redux, React Query, Axios 
- Redux Saga와 같은 녀석들은 라이브러리이긴 한데 Redux 미들웨어 역할을 수행하기 때문에 미들웨어라고도 불리우고 라이브러리라고도 불리웁니다.
